artifactory/commands/transferfiles/manager.go

```diff
@@ -207,23 +235,20 @@ func runProducerConsumers(pcWrapper *producerConsumerWrapper) (executionErr erro
 	go func() {
 		pcWrapper.chunkUploaderProducerConsumer.Run()
 	}()
-
 	go func() {
-		err := pcWrapper.chunkBuilderProducerConsumer.DoneWhenAllIdle(assumeProducerConsumerDoneWhenIdleForSeconds)
-		if err != nil {
-			log.Error("pcWrapper.chunkBuilderProducerConsumer.DoneWhenAllIdle API failed", err.Error())
-		}
+		// Wait till notified that the builder has no further tasks and close the builder producer consumer.
+		<-pcWrapper.builderFinishedNotifier
+		pcWrapper.chunkBuilderProducerConsumer.Done()
 	}()
 
 	// Run() is a blocking method, so once all chunk builders are idle, the tasks queue closes and Run() stops running.
 	pcWrapper.chunkBuilderProducerConsumer.Run()
-	// Once we are done building chunks, we can safely call DoneWhenAllIdle on the chunk uploader producer consumer.
-	// DoneWhenAllIdle is also a blocking method.
-	err := pcWrapper.chunkUploaderProducerConsumer.DoneWhenAllIdle(assumeProducerConsumerDoneWhenIdleForSeconds)
-	if err != nil {
-		log.Error("pcWrapper.chunkUploaderProducerConsumer.DoneWhenAllIdle API failed", err.Error())
+	if pcWrapper.chunkUploaderProducerConsumer.IsStarted() {
+		// Wait till notified that the uploader finished its tasks, and it will no longer receive further tasks from the builder.
```

Comment:
...and it will no longer receive further tasks from the builder.
-->
...and it will not receive new tasks from the builder.

Reply:
done